File: dsa/sparse_matrix/code/src/SpMatrix.js
const fs = require('fs').promises;
const path = require('path');
const readline = require('readline');

class SparseMatrix {
    constructor(rows, cols) {
        this.rows = rows;
        this.cols = cols;
        this.elements = new Map();
    }

    static async fromFile(filepath) {
        const fileContent = await fs.readFile(filepath, 'utf8');
        const lines = fileContent.split('\n');
        const rows = parseInt(lines[0].split('=')[1]);
        const cols = parseInt(lines[1].split('=')[1]);
        const matrix = new SparseMatrix(rows, cols);

        for (let i = 2; i < lines.length; i++) {
            const line = lines[i].trim();
            if (line) {
                const match = line.match(/\((\d+),\s*(\d+),\s*(-?\d+)\)/);
                if (match) {
                    const [, row, col, value] = match;
                    matrix.setElement(parseInt(row), parseInt(col), parseInt(value));
                } else {
                    throw new Error("Input file has wrong format");
                }
            }
        }
        return matrix;
    }

    getElement(row, col) {
        return this.elements.get(`${row},${col}`) || 0;
    }

    setElement(row, col, value) {
        if (row < 0 || row >= this.rows || col < 0 || col >= this.cols) {
            throw new Error("Invalid row or column index");
        }
        if (value !== 0) {
            this.elements.set(`${row},${col}`, value);
        } else {
            this.elements.delete(`${row},${col}`);
        }
    }

    add(other) {
        if (this.rows !== other.rows || this.cols !== other.cols) {
            throw new Error("Matrix dimensions do not match for addition");
        }
        const result = new SparseMatrix(this.rows, this.cols);
        for (let row = 0; row < this.rows; row++) {
            for (let col = 0; col < this.cols; col++) {
                const sum = this.getElement(row, col) + other.getElement(row, col);
                if (sum !== 0) {
                    result.setElement(row, col, sum);
                }
            }
        }
        return result;
    }

    subtract(other) {
        if (this.rows !== other.rows || this.cols !== other.cols) {
            throw new Error("Matrix dimensions do not match for subtraction");
        }
        const result = new SparseMatrix(this.rows, this.cols);
        for (let row = 0; row < this.rows; row++) {
            for (let col = 0; col < this.cols; col++) {
                const diff = this.getElement(row, col) - other.getElement(row, col);
                if (diff !== 0) {
                    result.setElement(row, col, diff);
                }
            }
        }
        return result;
    }

    multiply(other) {
        if (this.cols !== other.rows) {
            throw new Error("Matrix dimensions are not compatible for multiplication");
        }
        const result = new SparseMatrix(this.rows, other.cols);
        for (let row = 0; row < this.rows; row++) {
            for (let col = 0; col < other.cols; col++) {
                let sum = 0;
                for (let k = 0; k < this.cols; k++) {
                    sum += this.getElement(row, k) * other.getElement(k, col);
                }
                if (sum !== 0) {
                    result.setElement(row, col, sum);
                }
            }
        }
        return result;
    }

    async toFile(filepath) {
        let content = `rows=${this.rows}\ncols=${this.cols}\n`;
        for (const [key, value] of this.elements) {
            const [row, col] = key.split(',');
            content += `(${row}, ${col}, ${value})\n`;
        }
        await fs.writeFile(filepath, content);
    }
}

async function ensureDirectoryExists(dirPath) {
    try {
        await fs.access(dirPath);
    } catch (error) {
        await fs.mkdir(dirPath, { recursive: true });
    }
}

async function main() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    const question = (query) => new Promise((resolve) => rl.question(query, resolve));

    // Define paths according to the specified structure
    const baseDir = '/dsa/sparse_matrix'; // Base directory for the sparse matrix
    const inputDir = path.join(baseDir, 'sample_inputs'); // Input files directory
    const outputDir = path.join(baseDir, 'output'); // Output files directory

    // Ensure the output directory exists
    await ensureDirectoryExists(outputDir);

    try {
        while (true) {
            const operation = await question("Select operation (add/subtract/multiply) or 'quit' to exit: ");
            if (operation.toLowerCase() === 'quit') {
                break;
            }
            if (!['add', 'subtract', 'multiply'].includes(operation.toLowerCase())) {
                console.log("Invalid operation. Please try again.");
                continue;
            }

            const file1 = 'matrix1.txt'; // Specify the first input file directly
            const file2 = 'matrix2.txt'; // Specify the second input file directly
            const outputFile = `result_${operation}_${path.basename(file1)}_${path.basename(file2)}`;

            try {
                const matrix1 = await SparseMatrix.fromFile(path.join(inputDir, file1));
                const matrix2 = await SparseMatrix.fromFile(path.join(inputDir, file2));

                let result;
                switch (operation.toLowerCase()) {
                    case 'add':
                        result = matrix1.add(matrix2);
                        break;
                    case 'subtract':
                        result = matrix1.subtract(matrix2);
                        break;
                    case 'multiply':
                        result = matrix1.multiply(matrix2);
                        break;
                }

                const outputPath = path.join(outputDir, outputFile);
                await result.toFile(outputPath);
                console.log(`Operation completed. Result written to ${outputPath}`);
            } catch (error) {
                console.error("An error occurred:", error.message);
            }
        }
    } finally {
        rl.close();
    }
}

main().catch(console.error);

File: dsa/sparse_matrix/code/src/spmatrix.py
class SparseMatrix:
    def load_matrix(self, file_path):
        """
        Load the sparse matrix from a file.
        The file format contains the number of rows and columns followed by non-zero values.
        Each non-zero value is stored as (row, col, value).
        """
        matrix = {}
        try:
            # Open the specified file to read matrix dimensions
            with open(file_path, 'r') as file:
                rows = int(file.readline().split('=')[1])  # Get the total number of rows
                cols = int(file.readline().split('=')[1])  # Get the total number of columns

                self.numRows = rows
                self.numCols = cols

                # Process each line to read non-zero entries in the format (row, column, value)
                for line in file:
                    if line.strip():  # Ignore empty lines or whitespace
                        row, col, value = self.parse_entry(line)  # Parse the entry
                        if row not in matrix:
                            matrix[row] = {}  # Create a new row in the matrix if it does not exist
                        matrix[row][col] = value  # Store the value at the specified position

        except Exception as e:
            # Raise a Python standard exception for format issues
            raise ValueError(f"Input file has wrong format: {e}")
        
        return matrix  # Return the matrix that has been constructed

    def __init__(self, matrix_file=None, numRows=None, numCols=None):
        """
        Initialize the SparseMatrix class.
        - If `matrix_file` is provided, load the matrix from the file.
        - If `numRows` and `numCols` are provided, create an empty matrix with those dimensions.
        """
        if matrix_file:
            self.matrix = self.load_matrix(matrix_file)  # Load the matrix data from the specified file
            self.numRows = self.matrix.get(0, {}).get(0, 0)  # Get the number of rows from the loaded matrix
            self.numCols = self.matrix.get(0, {}).get(1, 0)  # Get the number of columns from the loaded matrix
        else:
            self.matrix = {}  # Initialize an empty matrix
            self.numRows = numRows
            self.numCols = numCols



    def parse_entry(self, line):
        """
        Parse a line containing a matrix entry.
        Each line is in the format (row, column, value).
        """
        # Remove parentheses and split the line by commas
        entry = line.strip()[1:-1].split(',')
        return int(entry[0]), int(entry[1]), int(entry[2])  # Return the parsed row, column, and value as integers

    def get_element(self, row, col):
        """
        Get the value of an element at a specific row and column.
        If the element is not explicitly stored (i.e., it's zero), return 0.
        """
        return self.matrix.get(row, {}).get(col, 0)  # Return the stored value or 0 if not found 

    def set_element(self, row, col, value):
        """
        Set a value at a specific row and column in the matrix.
        If the row or column doesn't exist, it will be initialized.
        """
        if row not in self.matrix:
            self.matrix[row] = {}  # Create a new dictionary for the row if it doesn't exist
        self.matrix[row][col] = value  # Assign the value at the given row and column

    def add(self, other_matrix):
        """
        Add this matrix to another sparse matrix.
        This operation adds non-zero elements and stores the result in a new matrix.
        """
        result = SparseMatrix(numRows=self.numRows, numCols=self.numCols)  # Create a result matrix

        # Loop through each element in the matrix
        for row in range(self.numRows):
            for col in range(self.numCols):
                # Sum the elements from both matrices at the current position
                result.set_element(row, col, self.get_element(row, col) + other_matrix.get_element(row, col))
        
        return result  # Return the matrix with the sum of elements

    def subtract(self, other_matrix):
        """
        Subtract another sparse matrix from this matrix.
        This operation subtracts non-zero elements and stores the result in a new matrix.
        """
        result = SparseMatrix(numRows=self.numRows, numCols=self.numCols)  # Create a result matrix

        # Iterate through every position in the matrix
        for row in range(self.numRows):
            for col in range(self.numCols):
                # Subtract corresponding elements from both matrices
                result.set_element(row, col, self.get_element(row, col) - other_matrix.get_element(row, col))
        
        return result  # Return the result of the subtraction

    def multiply(self, other_matrix):
        """
        Multiply this matrix with another sparse matrix.
        The operation follows matrix multiplication rules: The number of columns in the first matrix
        must be equal to the number of rows in the second matrix.
        """
        if self.numCols != other_matrix.numRows:
            raise ValueError("Matrix multiplication not possible with incompatible sizes.")  # Check matrix size compatibility

        result = SparseMatrix(numRows=self.numRows, numCols=other_matrix.numCols)  # Create a result matrix

        # Perform multiplication (row by column dot product)
        for row in self.matrix:
            for col in other_matrix.matrix:
                if col in self.matrix[row]:  # Only process non-zero elements
                    # Sum of product of corresponding elements
                    result.set_element(row, col, sum(self.get_element(row, k) * other_matrix.get_element(k, col) 
                                                     for k in range(self.numCols)))
        
        return result  # Return the result of the multiplication

    def save_to_file(self, file_path):
        """
        Save the sparse matrix to a file.
        The format will be similar to the input: number of rows, number of columns,
        followed by non-zero elements in (row, col, value) format.
        """
        with open(file_path, 'w') as file:
            # Write the number of rows and columns
            file.write(f"rows={self.numRows}\n")
            file.write(f"cols={self.numCols}\n")
            
            # Write each non-zero element in the matrix
            for row in self.matrix:
                for col, value in self.matrix[row].items():
                    file.write(f"({row}, {col}, {value})\n")


# Main execution block for user interaction
if __name__ == "__main__":
    # Initialize matrices from these file sources
    matrix1 = SparseMatrix(matrix_file='./easy_sample_02_3.txt')
    matrix2 = SparseMatrix(matrix_file='./easy_sample_03_2.txt')

    # Display options for matrix operations
    print("Select a matrix operation:")
    print("1. Matrix Addition")
    print("2. Matrix Subtraction")
    print("3. Matrix Multiplication")
    choice = int(input("Enter your choice (1/2/3): "))

    # Execute the chosen operation
    if choice == 1:
        result = matrix1.add(matrix2)
        print("The matrices have been successfully added!")
    elif choice == 2:
        result = matrix1.subtract(matrix2)
        print("The matrices have been successfully subtracted!")
    elif choice == 3:
        result = matrix1.multiply(matrix2)
        print("The matrices have been successfully multiplied!")
    else:
        print("Invalid input. Please select 1, 2, or 3.")

    # Ask the user to save the result to a file
    output_file = input("Please provide the file path where you'd like to save the result: ")
    result.save_to_file(output_file)